src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/PersonService.kt

```diff
@@ -0,0 +1,92 @@
+package uk.gov.justice.digital.hmpps.personrecord.service
+
+import org.springframework.stereotype.Service
+import org.springframework.transaction.annotation.Isolation
+import org.springframework.transaction.annotation.Transactional
+import uk.gov.justice.digital.hmpps.personrecord.jpa.entity.PersonAddressEntity
+import uk.gov.justice.digital.hmpps.personrecord.jpa.entity.PersonAliasEntity
+import uk.gov.justice.digital.hmpps.personrecord.jpa.entity.PersonContactEntity
+import uk.gov.justice.digital.hmpps.personrecord.jpa.entity.PersonEntity
+import uk.gov.justice.digital.hmpps.personrecord.jpa.repository.PersonRepository
+import uk.gov.justice.digital.hmpps.personrecord.model.Person
+import uk.gov.justice.digital.hmpps.personrecord.service.type.TelemetryEventType
+
+@Service
+class PersonService(
+  private val personRepository: PersonRepository,
+  private val telemetryService: TelemetryService,
+) {
+
+  @Transactional(isolation = Isolation.SERIALIZABLE)
+  fun processPerson(person: Person, callback: () -> PersonEntity?) {
+    val existingPersonEntity: PersonEntity? = callback()
+    handlePerson(person, existingPersonEntity)
+  }
+
+  private fun handlePerson(person: Person, existingPersonEntity: PersonEntity?) {
+    when {
+      (existingPersonEntity == null) -> handlePersonCreation(person)
+      else -> handlePersonUpdate(person, existingPersonEntity)
+    }
+  }
+
+  private fun handlePersonCreation(person: Person) {
+    createPersonEntity(person)
+    trackEvent(TelemetryEventType.CPR_RECORD_CREATED, mapOf("SourceSystem" to person.sourceSystemType.name))
+  }
+
+  private fun handlePersonUpdate(person: Person, existingPersonEntity: PersonEntity) {
+    updateExistingPersonEntity(person, existingPersonEntity)
+    trackEvent(TelemetryEventType.CPR_RECORD_UPDATED, mapOf("SourceSystem" to existingPersonEntity.sourceSystem.name))
+  }
+
+  private fun createPersonEntity(person: Person): PersonEntity {
+    val newPersonEntity = PersonEntity.from(person)
+    return updateAndSavePersonEntity(person, newPersonEntity)
+  }
+
+  private fun updateExistingPersonEntity(person: Person, personEntity: PersonEntity): PersonEntity {
+    var updatedPersonEntity = personEntity.update(person)
+    updatedPersonEntity = removeAllChildEntities(updatedPersonEntity)
```

Comment:
this is a bit confusing too. Do we genuinely want to do this? There are a lot of possible address sources and this will blow away any not coming from HMCTS (currently) as far as I understand it